storage/src/main/java/org/apache/kafka/storage/internals/log/LogCleaner.java

```diff
@@ -0,0 +1,652 @@
+ * higher offset than what is found in the segment (i.e. messages with a key that appears in the dirty section of the log).
+ * <p>
+ * To avoid segments shrinking to very small sizes with repeated cleanings we implement a rule by which if we will merge successive segments when
+ * doing a cleaning if their log and index size are less than the maximum log and index size prior to the clean beginning.
+ * <p>
+ * Cleaned segments are swapped into the log as they become available.
+ * <p>
+ * One nuance that the cleaner must handle is log truncation. If a log is truncated while it is being cleaned the cleaning of that log is aborted.
+ * <p>
+ * Messages with null payload are treated as deletes for the purpose of log compaction. This means that they receive special treatment by the cleaner.
+ * The cleaner will only retain delete records for a period of time to avoid accumulating space indefinitely. This period of time is configurable on a per-topic
+ * basis and is measured from the time the segment enters the clean portion of the log (at which point any prior message with that key has been removed).
+ * Delete markers in the clean section of the log that are older than this time will not be retained when log segments are being recopied as part of cleaning.
+ * This time is tracked by setting the base timestamp of a record batch with delete markers when the batch is recopied in the first cleaning that encounters
+ * it. The relative timestamps of the records in the batch are also modified when recopied in this cleaning according to the new base timestamp of the batch.
+ * <p>
+ * Note that cleaning is more complicated with the idempotent/transactional producer capabilities. The following
+ * are the key points:
+ * <p>
+ * <ol>
+ * <li>In order to maintain sequence number continuity for active producers, we always retain the last batch
+ *    from each producerId, even if all the records from the batch have been removed. The batch will be removed
+ *    once the producer either writes a new batch or is expired due to inactivity.</li>
+ * <li>We do not clean beyond the last stable offset. This ensures that all records observed by the cleaner have
+ *    been decided (i.e. committed or aborted). In particular, this allows us to use the transaction index to
+ *    collect the aborted transactions ahead of time.</li>
+ * <li>Records from aborted transactions are removed by the cleaner immediately without regard to record keys.</li>
+ * <li>Transaction markers are retained until all record batches from the same transaction have been removed and
+ *    a sufficient amount of time has passed to reasonably ensure that an active consumer wouldn't consume any
+ *    data from the transaction prior to reaching the offset of the marker. This follows the same logic used for
+ *    tombstone deletion.</li>
+ * </ol>
+ */
+public class LogCleaner implements BrokerReconfigurable {
+    private static final Logger LOG = LoggerFactory.getLogger(LogCleaner.class);
```

Comment:
`kafka.log.LogCleaner` is one of the few classes that has a logger defined in our default `log4j2.yaml` file. So we should either use the `LoggerFactory.getLogger()` method that takes a `String` and keep `kafka.log.LogCleaner` or update our `log4j2.yaml` file.

I'm unsure if changing the package can be considered a breaking change for this class.

Reply:
There's also instances in `Cleaner` and `CleanerThread`. 

As I said I'm unsure which approach to take, so I'd like to get input from other committers too.